Feliz.CompilerPlugins/ReactComponent.fs

```diff
@@ -0,0 +1,71 @@
+
+    /// <summary>Transforms call-site into createElement calls</summary>
+    override _.TransformCall(logger, memb, expr) =
+        let membArgs = memb.CurriedParameterGroups |> List.concat
+        match expr with
+        | Fable.Call(callee, info, typeInfo, range) when List.length membArgs = List.length info.Args ->
+            let callee =
+                match callee with
+                | Fable.Expr.IdentExpr ident -> Fable.Expr.IdentExpr { ident with Name = AstUtils.capitalize ident.Name }
+                | _ -> callee
+
+            if info.Args.Length = 1 && AstUtils.isRecord info.Args.[0].Type then
+                // F# Component { Value = 1 }
+                // JSX <Component Value={1} />
+                // JS createElement(Component, { Value: 1 })
+                AstUtils.makeCall (AstUtils.makeImport "createElement" "react") [callee; info.Args.[0]]
+            elif info.Args.Length = 1 && info.Args.[0].Type = Fable.Type.Unit then
+                // F# Component()
+                // JSX <Component />
+                // JS createElement(Component, null)
+                AstUtils.makeCall (AstUtils.makeImport "createElement" "react") [callee; AstUtils.nullValue]
+            else
+            let propsObj =
+                List.zip membArgs info.Args
+                |> List.choose (fun (arg, expr) -> arg.Name |> Option.map (fun k -> k, expr))
+                |> AstUtils.objExpr
+
+            // TODO: Detect usage of an input called `children` anc spread it into the component
+            AstUtils.makeCall (AstUtils.makeImport "createElement" "react") [callee; propsObj]
+        | _ -> expr
+
+    override this.Transform(logger, decl) =
+        if decl.Info.IsValue || decl.Info.IsGetter || decl.Info.IsSetter then
+            // Invalid attribute usage
+            let errorMessage = sprintf "Expecting a function declation for %s when using [<ReactComponent>]" decl.Name
+            logger.LogWarning(errorMessage, ?range=decl.Body.Range)
+            decl
+        else
+            // TODO: make sure isRecord works with records
+            if decl.Args.Length = 1 && AstUtils.isRecord decl.Args.[0].Type then
+                // do not rewrite components accepting records as input
+                { decl with Name = AstUtils.capitalize decl.Name }
+            else if decl.Args.Length = 1 && decl.Args.[0].Type = Fable.Type.Unit then
+                // remove arguments from functions requiring unit as input
+                { decl with Args = [ ]; Name = AstUtils.capitalize decl.Name }
+            else
+            // rewrite all other arguments into getters of a single props object
+            // TODO: transform any callback into into useCallback(callback) to stabilize reference
+            let propsArg = AstUtils.makeIdent (sprintf "%sProps" decl.Name)
+            let body =
+                ([], decl.Args) ||> List.fold (fun bindings arg ->
+                    // TODO: detect usage of "key" and emit warning
```

Comment:
I use this function to detect if an ident is being used in an expression (note you've to use `arg.Name` in case the name has been mangled): https://github.com/fable-compiler/Fable/blob/5690e334d53c0b985a5bbb4d6e141b305ceeec2b/src/Fable.Transforms/FableTransforms.fs#L165-L168

If we don't find `key` then we can just skip the `let key = $props.key` binding.

Reply:
Sounds like a plan :)